File: FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/LauncherActivity.java
/*
 * © 2016-2019 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.ui.activity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import com.flowcrypt.email.Constants;
import com.flowcrypt.email.R;
import com.flowcrypt.email.database.dao.source.AccountDao;
import com.flowcrypt.email.database.dao.source.AccountDaoSource;
import com.flowcrypt.email.database.dao.source.ActionQueueDaoSource;
import com.flowcrypt.email.jobscheduler.ForwardedAttachmentsDownloaderJobService;
import com.flowcrypt.email.jobscheduler.MessagesSenderJobService;
import com.flowcrypt.email.security.SecurityUtils;
import com.flowcrypt.email.service.EmailSyncService;
import com.flowcrypt.email.service.actionqueue.actions.EncryptPrivateKeysIfNeededAction;
import com.flowcrypt.email.ui.activity.base.BaseActivity;
import com.flowcrypt.email.util.SharedPreferencesHelper;

import androidx.preference.PreferenceManager;

/**
 * This is a launcher {@link Activity}
 *
 * @author Denis Bondarenko
 * Date: 3/5/19
 * Time: 9:57 AM
 * E-mail: DenBond7@gmail.com
 */
public class LauncherActivity extends BaseActivity {
  private AccountDao account;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    PreferenceManager.setDefaultValues(this, R.xml.preferences_notifications_settings, false);
    ForwardedAttachmentsDownloaderJobService.schedule(getApplicationContext());
    MessagesSenderJobService.schedule(getApplicationContext());

    account = new AccountDaoSource().getActiveAccountInformation(this);
    if (account != null && isNodeReady()) {
      showEmailManagerActivity();
    }
  }

  @Override
  protected void onNodeStateChanged(boolean isReady) {
    super.onNodeStateChanged(isReady);
    if (account != null) {
      showEmailManagerActivity();
    } else {
      showSignInActivity();
    }
  }

  @Override
  public boolean isDisplayHomeAsUpEnabled() {
    return false;
  }

  @Override
  public View getRootView() {
    return findViewById(R.id.layoutContent);
  }

  @Override
  public int getContentViewResourceId() {
    return R.layout.activity_launcher;
  }


  private void showSignInActivity() {
    startActivity(new Intent(this, SignInActivity.class));
    finish();
  }

  private void showEmailManagerActivity() {
    if (SecurityUtils.hasBackup(this)) {
      boolean isCheckKeysNeeded = SharedPreferencesHelper.getBoolean(PreferenceManager
          .getDefaultSharedPreferences(this), Constants.PREFERENCES_KEY_IS_CHECK_KEYS_NEEDED, true);

      if (isCheckKeysNeeded) {
        new ActionQueueDaoSource().addAction(this, new EncryptPrivateKeysIfNeededAction(account.getEmail()));
      }

      EmailSyncService.startEmailSyncService(this);
      EmailManagerActivity.runEmailManagerActivity(this);
      finish();
    }
  }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/LauncherActivity.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.ui.activity

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.preference.PreferenceManager
import com.flowcrypt.email.Constants
import com.flowcrypt.email.R
import com.flowcrypt.email.database.entity.AccountEntity
import com.flowcrypt.email.jetpack.viewmodel.LauncherViewModel
import com.flowcrypt.email.jetpack.workmanager.ForwardedAttachmentsDownloaderWorker
import com.flowcrypt.email.jetpack.workmanager.MessagesSenderWorker
import com.flowcrypt.email.service.FeedbackJobIntentService
import com.flowcrypt.email.service.IdleService
import com.flowcrypt.email.service.actionqueue.actions.EncryptPrivateKeysIfNeededAction
import com.flowcrypt.email.ui.activity.base.BaseActivity
import com.flowcrypt.email.util.CacheManager
import com.flowcrypt.email.util.FileAndDirectoryUtils
import com.flowcrypt.email.util.SharedPreferencesHelper

/**
 * This is a launcher [Activity]
 *
 * @author Denis Bondarenko
 * Date: 3/5/19
 * Time: 9:57 AM
 * E-mail: DenBond7@gmail.com
 */
class LauncherActivity : BaseActivity() {
  private val launcherViewModel: LauncherViewModel by viewModels()

  override val isDisplayHomeAsUpEnabled: Boolean
    get() = false

  override val rootView: View
    get() = findViewById(R.id.layoutContent)

  override val contentViewResourceId: Int
    get() = R.layout.activity_launcher

  override fun onCreate(savedInstanceState: Bundle?) {
    setupLauncherViewModel()
    super.onCreate(savedInstanceState)
    PreferenceManager.setDefaultValues(this, R.xml.preferences_notifications_settings, false)
    ForwardedAttachmentsDownloaderWorker.enqueue(applicationContext)
    MessagesSenderWorker.enqueue(applicationContext)
    FeedbackJobIntentService.enqueueWork(this)
    FileAndDirectoryUtils.cleanDir(CacheManager.getCurrentMsgTempDir())
  }

  override fun onAccountInfoRefreshed(accountEntity: AccountEntity?) {
    super.onAccountInfoRefreshed(accountEntity)
    launcherViewModel.isAccountInfoReceivedLiveData.value = true
  }

  private fun setupLauncherViewModel() {
    launcherViewModel.mediatorLiveData.observe(this, {
      if (launcherViewModel.isAccountInfoReceivedLiveData.value == true) {
        if (isAccountInfoReceived) {
          if (activeAccount != null) {
            showEmailManagerActivity()
          } else {
            showSignInActivity()
          }
        }
      }
    })
  }

  private fun showSignInActivity() {
    startActivity(Intent(this, SignInActivity::class.java))
    finish()
  }

  private fun showEmailManagerActivity() {
    val isCheckKeysNeeded = SharedPreferencesHelper.getBoolean(
      PreferenceManager
        .getDefaultSharedPreferences(this), Constants.PREF_KEY_IS_CHECK_KEYS_NEEDED, true
    )

    if (isCheckKeysNeeded) {
      roomBasicViewModel.addActionToQueue(
        EncryptPrivateKeysIfNeededAction(
          0,
          activeAccount!!.email,
          0
        )
      )
    }

    IdleService.start(this)
    EmailManagerActivity.runEmailManagerActivity(this)
    finish()
  }
}
